text/0036-unbounded-ints.md

```diff
@@ -0,0 +1,95 @@
+
+## Drawbacks
+
+### Performance:
+Using `i64`s is obviously faster than using unbounded integers. We benchmarked several crates, and the fastest we tested (under our licensing restrictions which forbids (L)GPL) was `IBig`. We evaluated policies that should represent the worst-case runtime increase from the switch.
+
+*Summary: In the worst case, we seem to be looking at 340μs vs 530μs for purely evaluation. This is a 56% increase, but negligible in an end-to-end system.*
+
+For the worst-case effect on policy evaluation, we assume a 10k character limit and do addition / multiplication of the literal `i64::MAX` repeated up to this limit +/* `principal.a` where `principal.a == i64::MAX`. All benchmarks are run for 100 iterations:
+
+```
+IBig_addition
+mean: 368.41585148514855μs
+std_dev: 37.14383484933371μs
+IBig_multiplication
+mean: 528.5201386138613μs
+std_dev: 27.376510998230426μs
+i128_addition
+mean: 341.99384158415853μs
+std_dev: 28.056390901018368μs
+i64_mult_by_1
+mean: 327.4614752475248μs
+std_dev: 28.18572847410058μs
+```
+
+Again, this is just evaluation. These differences are in the noise for end-to-end tests:
+
+```
+i128_10kchar_addition: 2.164s for 100 runs
+IBig_10kchar_addition: 2.165s for 100 runs
+```
+
+So an avarage run takes `~21ms` and the runtime increases by `~200μs`.
+
+### Applications Using Cedar Must Limit Input:
+In this RFC, we propose lifting Cedar's limits on integers. This means applications using Cedar would have to enforce their own limits. We expect applications to sanitize any user inputs and this probably includes limiting the lengths of string inputs. We think reasonable length limits (e.g., AVP limits policies to 10k chars) will give acceptable performance, but we should test this more (see unresolved question 2).
+
+### Taking new dependency:
+If this RFC is accepted, Cedar would take a new dependency on `IBig`, increasing the attack surface.
+
+## Alternatives
+
+### Unbounded internal integers, bounded inputs
+Applications will need to limit the integers input as attributes. Cedar could choose some default limit (e.g., `i64`). The current RFC doesn't impose a specific limit and relies on applications to sanitize any user inputs as appropriate.
+
+The downside of choosing a default limit is that analysis could return counter examples that don't respect this limit. It is not clear how to fix this while keeping analysis complete and decidable.
+
+It is probably sufficient to limit the length of the string passed as input, but this should be tested (see unresolved question 2).
+
+### Wrapping integers
+Arithmetic on integers that wrap on overflow is total, but it has caused bugs in the past and so is probably not ideal for an authorization language.
+
+## Unresolved questions
+
+- How efficient is (de)serializing `IBig`s?
+- Benchmarking was done assuming inputs bounded by `[i64::MIN, i64::MAX]` and 10k character policies. How does relaxing these assumptions affect performance?
```

Comment:
Does it make sense to limit the size of input integer values? Like this answer on stackoverflow: https://stackoverflow.com/a/39681707?

Reply:
> This proposal is mentioned above but has downsides. I responded in a comment; do you agree with it?

This would, again, complicate the analysis because we have to emit extra assertions for all integer fields that may be reachable in policy.  And the design then becomes inconsistent because ranges are allowed on fields but not in sets.  These are two disadvantages are addressed by the solution above, which is to express range constraints as-needed in policies instead.

Reply:
> These are two disadvantages are addressed by the solution above, which is to express range constraints as-needed in policies instead.

I don't believe they are fully addressed. If users really do intend that inputs to fall within a range, then they cannot express this fact, and counterexamples may still be outside the actual, but impossible-to-state, range. The solution above solves the problem of preventing overflow by pushing that responsibility on to the user to always check at run-time.

Reply:
> I don't believe they are fully addressed. If users really do intend that inputs to fall within a range, then they cannot express this fact, and counterexamples may still be outside the actual, but impossible-to-state, range. The solution above solves the problem of preventing overflow by pushing that responsibility on to the user to always check at run-time.

Checking range constraints in a program (or,  in our case, policy) is standard.  And there will always be constraints that aren't expressible in our type system (see [above](https://github.com/cedar-policy/rfcs/pull/36#discussion_r1380343683)).  So, I don't see this as a major problem.

Another advantage of policy-range constraints is that it's a two-way door.  If we choose to adopt BigInts at some point, all policies that were valid (i.e., not overflowing) under the existing semantics remain valid  in the BigInt semantics. 

But if we adopt BigInts now, there is no going back without a serious break in compatibility.  And, we will obligate all future implementations of Cedar to implement BigInt semantics.  This might carry a larger penalty in other host languages than it does in Rust.

To be clear, from the perspective of formalization and analysis, adopting unrestricted BigInts (so, BigInt literals and inputs are allows) is the easiest solution.   

But I'm not convinced that BigInts are the right long-term solution.  I hesitate to give up on a fundamental efficiency property of the language to more easily support what seems to be a rare use case (arithmetic in policies).